File: packages/java/src/main/java/dev/kreuzberg/HeadingLevel.java
package dev.kreuzberg;

import com.fasterxml.jackson.annotation.JsonCreator;
import com.fasterxml.jackson.annotation.JsonProperty;
import java.util.Objects;

/**
 * A single heading in the document hierarchy.
 */
public final class HeadingLevel {
	private final int level;
	private final String text;

	@JsonCreator
	public HeadingLevel(@JsonProperty("level") int level, @JsonProperty("text") String text) {
		this.level = level;
		this.text = Objects.requireNonNull(text, "text must not be null");
	}

	/**
	 * Get the heading depth (1 = h1, 2 = h2, etc.).
	 *
	 * @return heading depth
	 */
	public int getLevel() {
		return level;
	}

	/**
	 * Get the text content of the heading.
	 *
	 * @return heading text
	 */
	public String getText() {
		return text;
	}

	@Override
	public boolean equals(Object obj) {
		if (this == obj) {
			return true;
		}
		if (!(obj instanceof HeadingLevel)) {
			return false;
		}
		HeadingLevel other = (HeadingLevel) obj;
		return level == other.level && Objects.equals(text, other.text);
	}

	@Override
	public int hashCode() {
		return Objects.hash(level, text);
	}

	@Override
	public String toString() {
		return "HeadingLevel{level=" + level + ", text='" + text + "'}";
	}
}

File: packages/php/src/Types/HeadingLevel.php
<?php

declare(strict_types=1);

namespace Kreuzberg\Types;

/**
 * A single heading in the document hierarchy.
 *
 * @property-read int $level Heading depth (1 = h1, 2 = h2, etc.)
 * @property-read string $text Text content of the heading
 */
readonly class HeadingLevel
{
    public function __construct(
        public int $level,
        public string $text,
    ) {
    }

    /**
     * @param array<string, mixed> $data
     */
    /**
     * @param array<string, mixed> $data
     */
    public static function fromArray(array $data): self
    {
        /** @var int $level */
        $level = isset($data['level']) && is_int($data['level']) ? $data['level'] : 0;
        /** @var string $text */
        $text = isset($data['text']) && is_string($data['text']) ? $data['text'] : '';

        return new self(
            level: $level,
            text: $text,
        );
    }
}
